### PR TITLE
Port to Python 3

### PR DESCRIPTION
**Pull request by quozl**
Rebased from https://github.com/sugarlabs/sugar-datastore/pull/8.

Review invited from @tonadev and @chimosky.

Merge is planned once all critical components are ready.

### CODE REVIEW

File: src/carquinyol/layoutmanager.py [hunk 1]
```diff
@@ -104,6 +104,7 @@ def is_empty(self):
                         return False
         return True
 
+
```
Comment (Aniket21mathur) [inline src/carquinyol/layoutmanager.py]: Is this new line needed?

**Comment by Aniket21mathur** [comment]
I reviewed the code. Was able to built the package locally. What is the next step to take @quozl @pro-panda please guide.

**Comment by quozl** [comment]
Test it.  Remove the Python 2 installation of _sugar-datastore_, install from the Python 3 source, then start Sugar.  D-Bus should start the service.  Exercise the Journal.  This is what "_test running the datastore_" means in the tracking issue.

**Comment by Aniket21mathur** [comment]
I used sugar-toolkit-gtk3 @ _head_ , sugar @ _head_  to test the datastore, the datastore.logs file is not formed, and journal do not starts. Might be possible I am not setting up the testing environment correctly.
@pro-panda can you please tell how you tested it?

Thanks!

**Comment by quozl** [comment]
Datastore is a D-Bus service.  Just run it under a debugger?

**Comment by Aniket21mathur** [comment]
> Datastore is a D-Bus service. Just run it under a debugger?

Thanks. I am working on using `pdb` for debugging. Please suggest if there's anything better we can use ?

What exactly do we have to compare and watch through the debugger to ensure the correct working of sugar-datastore?

**Comment by quozl** [comment]
I'm not sure about better; it depends on what is needed, and the stack
is very deep.

I use gdb if there is an obvious terminating signal, such as a
segmentation fault, because that gives back control and you can make a
dump, backtrace or examine variables.  I use strace if I'm evaluating
which operating system calls (syscalls) are make.  I use dbus-monitor
to trace D-Bus API calls.  Others have used pdb for Python, ltrace for
library calls, and kernel call tracing for syscalls.

As baseline comparison, use the Python 2 build of the datastore.  It
should be nearly functionally identical.

For example, dbus-monitor output resulting from successful startup of
/usr/bin/datastore-service (Python 2) looks like this;

$ dbus-monitor --session
method call time=1562360626.708318 sender=:1.3092 -> destination=org.freedesktop.DBus serial=1 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=Hello
method return time=1562360626.708345 sender=org.freedesktop.DBus -> destination=:1.3092 serial=1 reply_serial=1
   string ":1.3092"
signal time=1562360626.708359 sender=org.freedesktop.DBus -> destination=(null destination) serial=4049 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string ":1.3092"
   string ""
   string ":1.3092"
signal time=1562360626.708381 sender=org.freedesktop.DBus -> destination=:1.3092 serial=2 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameAcquired
   string ":1.3092"
method call time=1562360626.708635 sender=:1.3092 -> destination=org.freedesktop.DBus serial=2 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=RequestName
   string "org.laptop.sugar.DataStore"
   uint32 0
signal time=1562360626.708664 sender=org.freedesktop.DBus -> destination=(null destination) serial=4050 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string "org.laptop.sugar.DataStore"
   string ""
   string ":1.3092"
signal time=1562360626.708686 sender=org.freedesktop.DBus -> destination=:1.3092 serial=3 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameAcquired
   string "org.laptop.sugar.DataStore"
method return time=1562360626.708698 sender=org.freedesktop.DBus -> destination=:1.3092 serial=4 reply_serial=2
   uint32 1
method call time=1562360626.725465 sender=:1.3092 -> destination=org.freedesktop.DBus serial=3 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=AddMatch
   string "type='signal',interface='org.freedesktop.DBus.Local',member='Disconnected'"
method return time=1562360626.725521 sender=org.freedesktop.DBus -> destination=:1.3092 serial=5 reply_serial=3

Above shows successful acquisition of the name
org.laptop.sugar.DataStore and the datastore program is running, ready
to respond to requests.  When the program is stopped with ctrl+c, the
name is released;

signal time=1562360741.364195 sender=:1.3092 -> destination=(null destination) serial=4 path=/org/laptop/sugar/DataStore; interface=org.laptop.sugar.DataStore; member=Stopped
signal time=1562360741.376864 sender=org.freedesktop.DBus -> destination=:1.3092 serial=5 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameLost
   string "org.laptop.sugar.DataStore"
signal time=1562360741.376949 sender=org.freedesktop.DBus -> destination=(null destination) serial=4051 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string "org.laptop.sugar.DataStore"
   string ":1.3092"
   string ""
signal time=1562360741.377054 sender=org.freedesktop.DBus -> destination=:1.3092 serial=6 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameLost
   string ":1.3092"
signal time=1562360741.377096 sender=org.freedesktop.DBus -> destination=(null destination) serial=4052 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string ":1.3092"
   string ":1.3092"
   string ""

Datastore can be tested in isolation without Sugar.  See
/usr/bin/copy-to-journal and /usr/bin/copy-from-journal

**Comment by Aniket21mathur** [comment]
Thanks, observed the output of the dbus monitor for python3 branch looks similar to python2.

**Comment by quozl** [comment]
Aniket Mathur wrote:
> Thanks, observed the output of the dbus monitor for python3 branch
> looks similar to python2.

Interesting, thanks.  It would be odd that datastore.log is not created and
the Journal does not start, yet the same D-Bus messages are seen.  You
will have to diagnose further; for example;

- identify the source code that creates datastore.log file, and verify
  it functions correctly after porting,

- examine the results of the datastore D-Bus API calls made by the
  Journal.

**Comment by quozl** [comment]
Tested https://github.com/sugarlabs/sugar-datastore/pull/14/commits/774d05f1fe53534ebf5b7b274a3f49e02a209bc5.  Fails with;

```log
dbus.exceptions.DBusException: org.freedesktop.DBus.Python.TypeError: Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/dbus/service.py", line 707, in _message_cb
    retval = candidate_method(self, *args, **keywords)
  File "/usr/local/lib/python3.6/dist-packages/carquinyol/datastore.py", line 308, in create
    self._metadata_store.store(uid, props)
  File "/usr/local/lib/python3.6/dist-packages/carquinyol/metadatastore.py", line 24, in store
    self._set_property(uid, key, value, md_path=metadata_path)
  File "/usr/local/lib/python3.6/dist-packages/carquinyol/metadatastore.py", line 51, in _set_property
    f.write(value)
TypeError: write() argument must be str, not dbus.Int32
```

My guess is the port needs more work.

**Comment by quozl** [comment]
https://dbus.freedesktop.org/doc/dbus-python/PY3PORT.html is a guide for porting D-Bus applications to Python 3.

**Comment by Aniket21mathur** [comment]
> My guess is the port needs more work.

Thanks. I am able to reproduce it, my configuration~
* Installed toolkit for python2 (HEAD).
* Using sugar (HEAD).
* sugar-artwork (HEAD).
* sugar-datastore `python3` branch.

Working to fix it.

**Comment by Aniket21mathur** [comment]
> https://dbus.freedesktop.org/doc/dbus-python/PY3PORT.html is a guide for porting D-Bus applications to Python 3.

Thanks!

**Comment by Aniket21mathur** [comment]
Replaced by #15

**Pull request closed without merging**